components/tooltip/tooltip.js

```diff
@@ -863,5 +881,31 @@ class Tooltip extends RtlMixin(LitElement) {
 		}
 		this._addListeners();
 	}
+
+	async _updateTruncating() {
+		// if no resize has happened since truncation was previously calculated the result will not have changed
+		if (!this._resizeRunSinceTruncationCheck || !this.onlyShowIfTruncating) return;
+
+		const target = this._target;
+
+		const clone = target.cloneNode(true);
+		clone.removeAttribute('id');
+		clone.style.position = 'absolute';
+		clone.style.overflow = 'hidden';
+		clone.style.whiteSpace = 'nowrap';
+		clone.style.width = '1px';
+
+		if (this.getAttribute('dir') === 'rtl') {
+			clone.style.right = '-10000px';
+		} else {
+			clone.style.left = '-10000px';
+		}
+
+		document.body.appendChild(clone);
+		await this.updateComplete;
+		this._truncating = clone.scrollWidth > target.offsetWidth;
```

Comment:
This calc assumes that the typography of the clone and the target will be the same.  However, consider...  

```html
<div class="d2l-body-small">
   <span id="the-id">...</span>
   <d2l-tooltip for="the-id" only-show-if-truncating>...</d2l-tooltip>
</div>
```
By appending the clone to the body, the clone measurement would be larger than if measured in the context of the target.  `font-weight`, `font-size`, and `letter-spacing` are the cascading properties I mostly worry about.  The heading classes would be cases where the target measurement would be bigger than the clone.

Previously the clone was being appended to the target, which I _think_ would have avoided this potential issue.  Did you run into troubles with it?  It's certainly possible to use `getComputedStyle` to mitigate this scenario, although the nice thing about appending to the target is that it factors in all the possible properties that might affect the width.  If the target happens to be a web component that only has named slots, or has some special behaviour associated slot changes, appending the the target may indeed be undesirable.  

I'm not sure if it will come up... it really depends on what the consumers do.

Reply:
> Previously the clone was being appended to the target, which I think would have avoided this potential issue. Did you run into troubles with it?

Yeah I ran into this issue early on with some examples I was using, which was why I had it appending to `target` instead - so that it would inherit the same font sizes and font styles.  I think we need to account for that issue, since I expect most of the time this will be used with small text sizes?

Reply:
I think when I played around with it, I had two elements...
```html
<div style="position: absolute; left: -10000px; width: 1px; overflow: hidden; whitespace: nowrap;">
   <the-clone></the-clone>
</div>
```
... and I measured the `scrollWidth` of `the-clone`.  Not sure if that's what you tried.

Reply:
I've now tried both ways and am getting the same `scrollWidth` of 60 still for both even though it is definitely not 60 - must be doing something different🤔

Reply:
I think it is getting this CSS from the demo page styles, which is preventing the button inside from growing.  I'm not entirely sure how it arrives at `60px` though.  If I set the clone's `max-width` to `none`, it gets the proper size. 
```css
.truncated d2l-button {
	margin-bottom: 20px;
	max-width: 100%;
}
```
In case you are interested... this is the code I tried...
```javascript
const cloneContainer = document.createElement('div');
cloneContainer.style.left = '-10000px';
cloneContainer.style.overflow = 'hidden';
cloneContainer.style.position = 'absolute';
cloneContainer.style.width = '1px';
cloneContainer.style.whiteSpace = 'nowrap';

const clone = target.cloneNode(true);
clone.removeAttribute('id');
clone.style.maxWidth = 'none';
cloneContainer.appendChild(clone);

if (this.getAttribute('dir') === 'rtl') {
	cloneContainer.style.right = '-10000px';
} else {
	cloneContainer.style.left = '-10000px';
}

target.appendChild(cloneContainer);
await this.updateComplete;
this._truncating = clone.scrollWidth > target.offsetWidth;
this._resizeRunSinceTruncationCheck = false;
target.removeChild(cloneContainer);
```
There's still the potential issue of dropping an element into a component's slot that it doesn't expect though.

Reply:
Ah that works for me as well! Thanks for digging into that.

That is a good point about the potential issue with putting an element into a component's slot - it could in theory result in some weird behaviour. In that case is do you think the safer option to put the clone on document and hope that `getComputedStyle` is enough to get an accurate measurement on the clone width, or hope that consumers have filters on any `slotchange` events/that this doesn't trigger any strange behaviours with slots? I tested with the subtle button slot and that seemed fine but I realize that's not a complex case with any `slotchange` listeners doing calculations.

Reply:
That's a good question - trade-offs either way.  

If the target is a web component that doesn't have a default (unnamed) slot, then the clone would not get rendered at all, but that scenario probably doesn't exist in our components today.  

What do we envision this being used for - buttons, dropdown openers maybe? bits of text, including headings? perhaps variants of tabs? Input elements cannot have children, but that case probably doesn't make sense anyway.  

I think I'd be inclined to append to the target, and keep the other sol'n in our back pocket - we can always pivot if we need to, and regressions are not a current risk since this is a new feature.  I'd be ok with either approach though.

Reply:
Yup I think those are along the lines for the potential use cases. Ok yeah I agree about appending to the target for now. I'll update that and add a comment in the code about the document alternative in case there end up being issues.